### PR TITLE
promtool: add rule test coverage reporting

### LINKED ISSUE #11848: promtool support for test and coverage of rules
### Proposal

We are using GitOps to write, validate and deploy Prometheus rules.

Currently, `promtool check config/rules` only writes to standard output and reports failure with return code.
This is enough for deployment gating  depending on the success/failure of the checks.

Concerning check of rules, only the code review can catch the case of a rule added without relevant test.
Being humans in a busy world, we can "miss" a rule and deploy the configuration without relevant test.

Would it be possible to have an output, understandable by a script, preferably using a well known format which would:

- report individual test success/error (that's for the nice display in Jenkins by example)
- report coverage of rules in a file (much harder)

I recognize that the coverage request is opening a (small) can of worm  but it is the most interesting one IMHO.

Note: I think a minimal xunit/junit format would be a good option since it is widely supported

Example: 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
    <!-- test suite could correspond to test_group  -->
    <testsuite timestamp="2023-01-13T10:05:06">
         <!-- test_group name for "className"  -->
         <!-- for name of case, nothing exists ; index in suite may be enough - for alert_test_case, alertame coube part of it. -->
        <testcase name="???" classname="test_group.name" time="0.001">
            <failure message="FAILED alertname:..." type="result.DoesntMatch">
                <!-- message -->
                exp:"[....]"
                got:"[...]"
            </failure>
        </testcase>
    </testsuite>
</testsuites>
```

Coverage would be more tricky. For my use case, I would settle for a flag generating errors is a rule hasn't been tested in a run (could be an informational entry in output).

### PR DESCRIPTION
**Pull request by thc1006**
This patch adds `--coverage` and `--coverage-threshold` to `promtool test rules`. `--coverage` reports which alerting and recording rules of the loaded rule files are exercised by unit test assertions, and `--coverage-threshold` fails the command when total coverage is below a percentage, so a CI pipeline can refuse a rule that arrived without a test.

In a GitOps repository nothing but code review notices a rule landing untested, and code review is exactly what gets skipped when an alert is urgent. This is the coverage half of #11848; the JUnit half was already resolved by #14506.

```console
$ promtool test rules --coverage coverage_test.yml
  SUCCESS

Rule test coverage:
  Alerting rules:  2/3 covered (66.7%)
  Recording rules: 1/2 covered (50.0%)
  Total:           3/5 covered (60.0%)
  Uncovered rules:
    group "alerts" in coverage_rules.yml:
      - alert: NeverTested{severity="warning"} (rule #3)
    group "recording" in coverage_rules.yml:
      - record: job:up:avg (rule #2)
```

```console
$ promtool test rules --coverage-threshold=80 coverage_test.yml
...
Rule test coverage 3/5 (60.0%) is below the threshold of 80%.
$ echo $?
1
```

## What counts as covered

Coverage here is assertion presence, not expression branch coverage. It answers "does any test assert on this rule's output?", not "does every branch of its expression get evaluated". A rule is covered when an `alert_rule_test` names it in `alertname`, or a `promql_expr_test` contains a selector that could read its output.

A failing assertion still counts as covered, because the rule is under test; the failure fails the command on its own and does not need to be counted twice. Because promtool evaluates same-named alerts together and compares against their unioned output, an `alert_rule_test` covers every rule with that `alertname` — that is a property of the assertion, not an approximation on my side.

All the test files on the command line form one suite, so a rule file shared by several of them is counted once rather than once per referencing file.

This is deliberately narrower than the input-feeding model sketched in #8131, where a rule is covered if the test's `input_series` supply its inputs. Asking about the output puts the question closer to what a test is for than asking whether the rule ran, though neither model establishes that the assertion observed this rule's output; the limits section below says what this one does not check. I would rather ship the narrower thing and let #8131 stay open for the dependency-graph work.

## How rules are identified

Rules come out of `rules.Manager.LoadGroups` rather than a hand-rolled `rulefmt` walk, so group and rule labels are merged exactly the way Prometheus merges them.

Each rule is keyed by its rule file, its group, and its position within that group. Position rather than content matters here: two declarations in one group can share a name, an expression and a set of labels and still be two rules, differing only in something coverage does not track such as `for`. Keying on content would silently merge them and quietly shrink the denominator. Rule file paths are resolved through symlinks first, so the same physical file reached by two routes is still one file, and the report prints enough of each path to tell apart rule files that share a base name.

## Selector attribution

For recording rules the selector's `__name__` matchers have to accept the rule's output metric; for alerting rules the selector has to be able to read the rule's `ALERTS` or `ALERTS_FOR_STATE` series with a matching `alertname`.

The matching follows what the querying documentation promises:

> Multiple matchers can be used for the same label name; they all must pass for a result to be returned.

and

> Label matchers can also be applied to metric names by matching against the internal `__name__` label.

so `{__name__=~"job:up:sum|other",__name__!="job:up:sum"}` is not attributed to `job:up:sum`, and neither is `ALERTS{alertname=~"A|B",alertname!="A"}` attributed to `A`. Labels the rule engine generates are compared against what the engine actually produces rather than against the rule's own labels: `alertname` is always the rule name, and on `ALERTS_FOR_STATE` there is no generated `alertstate` at all, so there it is an ordinary label. `alertname` is likewise ordinary on a recording rule's output and is compared like any other label.

`alertstate` is matched against the states the rule can reach rather than against both of them. An alert is only sampled once its state has settled for the evaluation, and a rule with no `for` is promoted to firing before its first sample, so `ALERTS{alertname="X",alertstate="pending"}` observes nothing for such a rule and does not cover it.

Alert labels are expanded per alert instance, so a value containing template syntax is treated as unknown instead of compared literally. `AlertingRule.QueryForStateSeries` already skips those labels when it builds matchers, and it seemed wrong for coverage to be stricter about them than the rule engine is.

## Making the threshold safe to gate on

Two guarantees are worth keeping apart, because only one of them is sound.

Deciding which rules exist fails closed, and the threshold arithmetic is exact. It is compared against the raw covered and total counts and never against the percentage as it is rendered; rounding first is not hypothetical, since 1999 of 2000 rules is 99.95%, prints as `100.0%`, and would clear a threshold of 100 with a rule still untested.

Deciding which rules an assertion covers is the heuristic above, and can over-report. A passing threshold means every rule was attributed an assertion by this model, not that every rule was proven to be observed by one. The documentation says so in those words rather than promising more.

One limit is worth stating before the rest, because it decides whether the gate does the job #11848 asks of it. The denominator holds the rules of the rule files the given test files reach through their `rule_files`, and nothing else. A rule file nobody wrote a test file for is not in the suite: no warning, no total, no effect on the percentage. A rule added without a test is caught; a rule *file* added without a test file is not, unless `rule_files` globs the production tree or the test files are generated from the same manifest. The documentation says this where someone setting up the gate will read it.

Coverage that could not be determined does not pass either, and there are three ways to not determine it. A rule file that fails to load never reaches the denominator, and with no test group to run nothing else would load it, so the report would come out `0/0` and be read as fully covered. A `rule_files` entry that matches nothing is worse, because it disappears quietly: the files that do exist can be fully covered and the total is not even zero. A suite with no rules at all is the same question with no data. All three fail a positive threshold with the reason printed.

Report-only mode keeps its exit code and only warns, since `--coverage` has always been informational and the test run reports rule file errors of its own. A threshold failure is also recorded in the JUnit XML as a failing `coverage threshold` test case, so a run that fails only on coverage does not read as green to whatever consumes that file.

Non-finite thresholds are rejected up front. `strconv.ParseFloat` accepts `NaN`, every comparison against `NaN` is false, and an unvalidated `NaN` would sail through the range check and then quietly disable the very gate it was asked to enforce.

## Where it is imprecise

Attribution is a static analysis of the selectors, so it misses in both directions and the documentation says so rather than promising one.

It under-reports when a selector has no concrete `__name__`, or when an alert selector has no `alertname`: neither can be attributed to a particular rule, so a test that genuinely exercises one that way is not counted. #8131 notes the same limitation.

It over-reports because it reasons about what a selector *could* read rather than what it did. A matcher on a label the rule does not set counts when some value could satisfy every matcher on it, whether or not the rule ever emits that value: `job:up:sum{job="prod"}` counts even for a rule that only ever emits `job="staging"`. Templated alert labels are likewise treated as unknown rather than expanded.

What it does reject is a matcher set that no value can satisfy, and it rejects those by finding a value rather than by pattern-matching the contradiction. Candidates come from the matchers themselves, from the literal alternations `SetMatches` exposes, and from strings built by walking the parsed expression: every alternation branch, a bounded set of representative character-class endpoints, and a bounded product across concatenations and repetitions, capped so a nested alternation cannot blow the search up. Parsing uses the flags `labels.NewFastRegexMatcher` uses, so `.` means the same thing here as it does when the matcher runs.

`{job=~"a.*",job!~"a.*"}` and `{job=~"a.*",job=~"b.*"}` have no satisfying value and do not count. `{job=~"(staging|prod)-.*",job!~"staging-.*"}` and `{job=~"[ab]+",job!~"a+"}` do, via `prod-` and `b`, which is why the search has to try more than the first branch it finds.

There are three outcomes, not two. A value is found and the rule counts. The matchers confine the label to a finite set, every member is tried, and none works, which rules the selector out. Or the search runs out of candidates or budget, and the answer is that it does not know. That last case used to be reported as uncovered, which told the user something the analysis had not established: `{job=~"a+",job!~"a|aa|aaa"}` reads a rule through `aaaa`, and the search never builds that value. Undecided rules are now listed separately, and the gate says the coverage lies between two bounds rather than naming them untested. It still fails, so nothing opens up.

These really are candidates rather than witnesses. A zero-width assertion can contradict the parts around it, so the walk builds `foobar` for `foo\bbar` even though that expression matches nothing. Every candidate is confirmed against the real matchers before it counts, which is what keeps a badly built one costing a missed attribution rather than a wrong one.

The search being bounded is where the under-reporting lives: intersecting two unbounded expressions needs a value neither describes on its own, as in `{job=~".*foo",job=~"xyz.*"}`, which `xyzfoo` satisfies. A brute-force test over a small alphabet sweeps pairs of regexp matchers and asserts the search finds a value whenever one exists. That pair is the only gap in that table, which is a finite table rather than a proof: a matcher set whose solutions all sit outside the candidates the search builds will still be missed.

It also matches on selectors alone, and deliberately establishes neither where the samples an assertion sees come from nor when they exist. `input_series` supplying `ALERTS` or a recorded metric directly counts as covering the rule of that name; `eval_time`, `offset` and `@` are not considered; and a selector that does not affect the result, as in `r or vector(1)`, still counts. Sample provenance and rule dependencies are the broader problem #8131 describes, and are out of scope here.

## Note on this revision

The branch is rebased onto current `main` and force-pushed. It is still a single commit.

Since the last review, three more ways the gate could pass without observing a rule are closed. `alertstate` was assumed to reach `pending` for every alerting rule, so an assertion on the pending state of a rule with no `for` counted as coverage while observing nothing. An unbounded regexp made the whole matcher set compatible, so `{job=~"a.*",job!~"a.*"}` counted too. In the other direction, a matcher set that only excluded values was decided against a handful of fixed probes, so `{job!="",job!="a",job!="A",job!="0",job!="-"}` was wrongly called unsatisfiable although `aa` satisfies it.

Both threshold regressions for the first two go through `RulesUnitTestResult` rather than the helpers, since that is where the fail-open showed.

Attribution on dynamic labels is witness-based: a positive result always has a concrete value that satisfies every matcher. The search enumerates rather than taking the first path through the expression, since a selector like `{job=~"(staging|prod)-.*",job!~"staging-.*"}` is perfectly ordinary and its only witness is in the second branch.

The solver reports three outcomes rather than two, and the candidate search is bounded and cached per selector; the sections above describe both. Before this, an exhausted search was indistinguishable from a proven contradiction, and a legal `[a-z]{1000}` matcher allocated 33.6 MiB for every rule it was compared against.

One false green is closed. Stripping templated labels rewrote the rule's label set through `labels.Builder`, which deletes labels whose value is the empty string, so a rule setting `severity: ""` alongside a templated label lost the `severity` constraint entirely and `ALERTS{alertname="X",severity="page"}` counted as covering it. Nothing is rewritten now; whether a value is known is decided where it is compared.

Three things changed in the round before. Two bounded-search misses are closed: a negative-only matcher set was decided against the probes and a single fresh-value family, so `{job!~"|a|A|0|-",job!~"coverage-witness-.*"}` was called unsatisfiable although `b` satisfies it, and repetitions were only built one and two deep, so `{job=~"a+",job!~"a|aa"}` missed `aaa`. Neither is a completeness proof, only a wider net; the wording above says so now.

Also, `\B` was grouped with the impossible-node case, so `{job=~"foo\Bbar"}` was read as unsatisfiable although it matches `foobar`; it is an ordinary zero-width assertion and is treated like the others now. The uncovered list also prints each rule's position in its group, because two declarations differing only in `for` produced two identical lines and no way to tell which one needed the test:

```
  Uncovered rules:
    group "partial" in rules.yml:
      - alert: Twin (rule #1)
```

The two end-to-end threshold tests for this now assert on a real sample rather than an empty result, so they show the selector reading the rule rather than only being attributable to it.

The coverage code and its tests have moved to `rule_coverage.go` and `rule_coverage_test.go`. Nothing changed with them; `unittest.go` is down to the flag plumbing and one call, which should make the rest of this easier to read.

On cost, the walk is bounded by a byte budget and each selector keeps its own answers, since the answer for a label depends on the matchers rather than on the rule being compared. Both were needed: `[a-z]{1000}` is a legal matcher that expands into a thousand-element concatenation, and it allocated 33.6 MiB per call, once per rule the selector was compared against. It is 0.23 MiB now, and a selector compared against 200 rules solves once, so those 200 rules cost 376µs in total rather than 200 times 325µs. An equality matcher is 92ns. A suite of 200 rules against 20 such selectors runs in 0.09s.

Earlier in this revision, two ways the gate could pass without complete coverage were closed. A `rule_files` entry that matched nothing used to warn and then vanish, so a suite whose remaining files were fully covered satisfied a threshold of 100 while part of it had never been analysed; unmatched patterns are now carried through and counted as unanalysed. Separately, satisfiability of a dynamic label was only checked when an equality matcher pinned it, so regexp-only contradictions such as `{job=~"a",job!~"a"}` and `{job!~".*"}` were accepted and an assertion that can never observe a rule marked it covered.

For the unmatched-pattern part I deliberately did not reintroduce `--warn-as-error`: #16932 showed that turning warnings into an API contract wants a design covering query-engine warnings, warning types and exit codes. This only lets the coverage gate see what it was not given, and leaves the broader question in #12499 open.

Also in this revision: the documentation no longer claims attribution can only under-report, which was not true of templated labels and is not true of the regexp cases either; failed rule files are counted per file rather than per error; and a threshold failure now appears in the JUnit XML.

The coverage-only rule manager passes a logger explicitly. Without one it inherited a nil logger from `NewManager` and panicked on any rule file the loader warns about, which is a bug in the rule manager rather than in this feature; it is filed as #19432 with a fix in #19433, and this PR does not depend on that one.

## Testing

`go test ./cmd/promtool/...` passes, including with `-race`, under `GOARCH=386`, and with `-race` under each of the `stringlabels`, `dedupelabels` and `slicelabels` build tags; `golangci-lint run cmd/promtool/...` is clean and the CLI documentation is regenerated. 122 coverage and selector tests and subtests pass in total.

One note on CI: `slicelabels` runs over `./cmd/prometheus ./model/textparse ./prompb/...` rather than `./...`, so this package is not covered by that job. I ran it locally instead and left the workflow alone here. #19447 adds `./cmd/promtool/...` to that line on its own, and this PR does not depend on it.

The regression tests cover the behaviour this revision changes: an alert with and without `for` against a `pending` selector, an unbounded regexp required and excluded, two required regexps with disjoint literal prefixes, nested prefixes and a regexp without one that both have to stay compatible, and a negative-only matcher set excluding every fixed probe. Two of these go end to end through `RulesUnitTestResult` with a threshold rather than through the helpers. Earlier ones cover a valid fully covered rule file alongside one that matches nothing, contradictory finite regexp matchers, a negated total regexp, a threshold failure appearing in the JUnit output, and a rule file the loader warns about, an unloadable rule file with no test group and with `--run` selecting none, a suite with no rules, `NaN` and the infinities, 1999 of 2000 rules against a threshold of 100, repeated `__name__` and `alertname` matchers, a static `alertname` on a recording rule, a templated alert label, an impossible `alertstate`, two lookalike declarations in one group, rule files sharing a base name, a rule file reached through a symlink, and a rule file shared by two test files.

Where a test guards a fix rather than a feature I checked that it fails without the fix.

#### Which issue(s) does the PR fix:

Fixes #11848

Related to #8131 — this covers the output-assertion part of it, not the input and dependency analysis discussed there.

#### Release notes for end users (**ALL** commits must be considered).

```release-notes
[FEATURE] promtool: Add `--coverage` and `--coverage-threshold` flags to `promtool test rules` to report which alerting and recording rules are exercised by unit test assertions and optionally fail when coverage is below a threshold.
```

### CODE REVIEW

File: cmd/promtool/unittest.go [hunk 1]
```diff
@@ -736,3 +770,177 @@ func (ps *parsedSample) String() string {
...
+	if err := resolveAndGlobFilepaths(filepath.Dir(testFile), &utf); err != nil {
+		return nil
+	}
+
+	p := parser.NewParser(parser.Options{})
+	var entries []ruleEntry
+	for _, rf := range utf.RuleFiles {
+		rgs, errs := rulefmt.ParseFile(rf, true, model.UTF8Validation, p)
+		if len(errs) > 0 || rgs == nil {
+			continue
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Fixed. `loadRuleEntries` now receives the caller's `parser.Parser` and `ignoreUnknownFields` flag, and passes them through to `rulefmt.ParseFile`.

File: cmd/promtool/unittest.go [hunk 2]
```diff
@@ -736,3 +770,177 @@ func (ps *parsedSample) String() string {
...
+		}
+		for _, group := range rgs.Groups {
+			for _, rule := range group.Rules {
+				var e ruleEntry
+				e.group = group.Name
+				e.file = rf
+				e.labels = labels.FromMap(rule.Labels)
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Fixed. Now using `rules.FromMaps(group.Labels, rule.Labels)` to merge group-level and rule-level labels, matching the precedence in `rules.Manager.LoadGroups`.

File: cmd/promtool/unittest.go [hunk 3]
```diff
@@ -736,3 +770,177 @@ func (ps *parsedSample) String() string {
...
+		for _, m := range matchers {
+			if m.Name == labels.AlertName {
+				return m.Matches(entry.name)
+			}
+		}
+		// No alertname matcher — indeterminate, don't mark as covered.
+		return false
+	}
+
+	// Direct name match for recording rules (and alert rules by name).
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Fixed. Added an early `return false` guard for alert rules before the direct `__name__` match. Alert rules can now only be covered via `ALERTS{alertname=...}` selectors or `alert_rule_test`. Added test case `alert rule not matched by direct __name__`.

File: cmd/promtool/unittest.go [hunk 4]
```diff
@@ -736,3 +770,177 @@ func (ps *parsedSample) String() string {
...
+	for _, m := range matchers {
+		if m.Name == labels.MetricName {
+			continue
+		}
+		if v := entry.labels.Get(m.Name); v != "" {
+			if !m.Matches(v) {
+				return false
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Fixed. Replaced `Get() != ""` with `entry.labels.Has(m.Name)` to correctly distinguish between an empty-value label and a missing label.

File: cmd/promtool/unittest.go [hunk 5]
```diff
@@ -736,3 +770,177 @@ func (ps *parsedSample) String() string {
...
+			covered++
+		}
+	}
+
+	pct := 0.0
+	if total > 0 {
+		pct = float64(covered) / float64(total) * 100
+	}
+	fmt.Fprintf(os.Stderr, "Rule test coverage: %d/%d rules covered (%.1f%%)\n", covered, total, pct)
+
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Intentional. The existing test runner prints `SUCCESS`/`FAILED` to stdout and debug/error output to stderr. Coverage is an informational supplement — printing it to stdout would interleave with the test result lines and break existing output parsing. This is consistent with how `--debug` output is handled.

File: cmd/promtool/unittest.go [hunk 6]
```diff
@@ -72,18 +80,22 @@ func RulesUnitTestResult(results io.Writer, queryOpts promqltest.LazyLoaderOpts,
 			fmt.Println("  SUCCESS")
 		}
 		fmt.Println()
+		allRuleEntries = append(allRuleEntries, entries...)
 	}
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Acknowledged. In practice `promtool test rules` is called with one test file at a time, and different test files typically reference different rule files. Adding deduplication logic increases complexity for an edge case that is unlikely in real workflows. Happy to add it in a follow-up if maintainers consider it important.

File: cmd/promtool/unittest.go [hunk 7]
```diff
@@ -736,3 +770,184 @@ func (ps *parsedSample) String() string {
...
+// are merged with rule-level labels, matching the behaviour of rules.Manager.LoadGroups.
+func loadRuleEntries(testFile string, p parser.Parser, ignoreUnknownFields bool) []ruleEntry {
+	content, err := os.ReadFile(testFile)
+	if err != nil {
+		return nil
+	}
+
+	var utf unitTestFile
+	if err := yaml.UnmarshalStrict(content, &utf); err != nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: No change needed. The existing test runner (`ruleUnitTest` at line 106) also uses `yaml.UnmarshalStrict` to parse the test YAML file. The `--ignore-unknown-fields` flag only applies to rule files (passed to `m.LoadGroups`), not to the test file itself. Our behaviour is consistent with the existing code path.

File: cmd/promtool/unittest.go [hunk 8]
```diff
@@ -736,3 +770,184 @@ func (ps *parsedSample) String() string {
...
+	if err := resolveAndGlobFilepaths(filepath.Dir(testFile), &utf); err != nil {
+		return nil
+	}
+
+	var entries []ruleEntry
+	for _, rf := range utf.RuleFiles {
+		rgs, errs := rulefmt.ParseFile(rf, ignoreUnknownFields, model.UTF8Validation, p)
+		if len(errs) > 0 || rgs == nil {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Fixed. All error paths in `loadRuleEntries` now emit a `warning: coverage: ...` message to stderr before returning nil or continuing. This lets users know when coverage data may be incomplete without failing the test run.

File: cmd/promtool/unittest.go [hunk 9]
```diff
@@ -736,3 +770,184 @@ func (ps *parsedSample) String() string {
...
+			return false
+		}
+		// Look for an alertname matcher.
+		for _, m := range matchers {
+			if m.Name == labels.AlertName {
+				return m.Matches(entry.name)
+			}
+		}
+		// No alertname matcher — indeterminate, don't mark as covered.
+		return false
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Fixed. The ALERTS/ALERTS_FOR_STATE path now checks all non-`__name__`/non-`alertname` matchers against the alert rule's static labels using the same `Has()`/`Get()` pattern as recording rules. Added test cases `ALERTS_with_compatible_static_labels` and `ALERTS_with_incompatible_static_labels`.

File: cmd/promtool/unittest.go [hunk 10]
```diff
@@ -72,18 +80,22 @@ func RulesUnitTestResult(results io.Writer, queryOpts promqltest.LazyLoaderOpts,
...
 		fmt.Println()
+		allRuleEntries = append(allRuleEntries, entries...)
 	}
 	err := junit.WriteXML(results)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "failed to write JUnit XML: %s\n", err)
 	}
+	if coverage && len(allRuleEntries) > 0 {
+		printCoverageSummary(allRuleEntries)
+	}
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest.go]: Fixed. `TestRulesUnitTestCoverage` now calls `RulesUnitTestResult` with `coverage=true` as an end-to-end integration test, exercising the full CLI path including rule entry loading, coverage marking during test execution, and summary aggregation.

File: docs/command-line/promtool.md [hunk 11]
```diff
@@ -477,6 +477,7 @@ Unit tests for rules.
 | <code class="text-nowrap">--debug</code> | Enable unit test debugging. | `false` |
 | <code class="text-nowrap">--diff</code> | [Experimental] Print colored differential output between expected & received output. | `false` |
 | <code class="text-nowrap">--ignore-unknown-fields</code> | Ignore unknown fields in the test files. This is useful when you want to extend rule files with custom metadata. Ensure that those fields are removed before loading them into the Prometheus server as it performs strict checks by default. | `false` |
+| <code class="text-nowrap">--coverage</code> | Report rule test coverage after running tests. | `false` |
```
[the remark this answers is not in the document]
Comment (thc1006) [reply docs/command-line/promtool.md]: Fixed. Updated the flag description in main.go to "Report rule test coverage to stderr after running tests." and regenerated docs/command-line/promtool.md.

File: cmd/promtool/unittest_test.go [hunk 12]
```diff
@@ -198,6 +199,165 @@ func TestRulesUnitTest(t *testing.T) {
...
+
+	// End-to-end: run RulesUnitTestResult with coverage=true through the full test runner.
+	var buf bytes.Buffer
+	got := RulesUnitTestResult(&buf, promqltest.LazyLoaderOpts{}, parser.NewParser(parser.Options{}), nil, false, false, false, true, "./testdata/coverage_test.yml")
+	require.Equal(t, 0, got, "expected tests to pass with coverage enabled")
+
+	// Verify coverage tracking via helpers.
+	entries := loadRuleEntries("./testdata/coverage_test.yml", parser.NewParser(parser.Options{}), false)
+	require.Len(t, entries, 5)
+
```
[the remark this answers is not in the document]
Comment (thc1006) [reply cmd/promtool/unittest_test.go]: Simplified the test. `TestRulesUnitTestCoverage` now runs the full `RulesUnitTestResult(coverage=true)` end-to-end and verifies return code, then validates `loadRuleEntries` correctly discovers all 5 rules (3 alerts + 2 recordings). The detailed selector matching logic is covered by `TestRuleMatchesSelector` with 14 table-driven cases. Capturing stderr for summary assertion would require non-parallel-safe pipe tricks — the current split between integration (e2e return code) and unit (selector matching) tests provides sufficient coverage without sacrificing test parallelism.

File: cmd/promtool/unittest.go [hunk 13]
```diff
@@ -736,3 +770,210 @@ func (ps *parsedSample) String() string {
 	}
 	return ps.Labels.String() + " " + strconv.FormatFloat(ps.Value, 'E', -1, 64)
 }
+
+// ruleEntry tracks a single rule for coverage analysis.
+type ruleEntry struct {
+	name    string
+	group   string
+	file    string
+	rtype   string // "alert" or "record"
```
Comment (krajorama) [inline cmd/promtool/unittest.go]: Let's use a proper type and constants instead of copy paste strings please, as in:
```
type ruleKind uint8

const (
	alertingRule ruleKind = iota
	recordingRule
)

type ruleEntry struct {
	name    string
	group   string
	file    string
	kind    ruleKind
	labels  labels.Labels
	covered bool
}
```

File: cmd/promtool/unittest.go [hunk 14]
```diff
@@ -736,3 +770,210 @@ func (ps *parsedSample) String() string {
...
+	labels  labels.Labels
+	covered bool
+}
+
+// loadRuleEntries reads rule files referenced by a unit test file and returns
+// a ruleEntry for each alerting and recording rule found. Group-level labels
+// are merged with rule-level labels, matching the behaviour of rules.Manager.LoadGroups.
+func loadRuleEntries(testFile string, p parser.Parser, ignoreUnknownFields bool) []ruleEntry {
```
Comment (krajorama) [inline cmd/promtool/unittest.go]: This is a duplicate of `LoadGroups`. Let's not re-implement that. We should instead process the output of `LoadGroups`. Especially since it solves two bugs with this implementation.

1. `loadRuleEntries` is called once per *test file* (unittest.go:67-70) and each result is
appended to `allRuleEntries`, matched only against that file's own tests. A rule file
referenced by N test files is counted N times.

Repro: two test files both reference a `rules2.yml` with a single recording rule
`only:metric`; only the first tests it.

```
Rule test coverage: 1/2 rules covered (50.0%)
  Untested rules:
    group "g" in rules2.yml:
      - record: only:metric
```

The same physical rule appears as **both covered and untested**, denominator inflated.
This is exactly the GitOps scenario the feature targets (shared rule files, multiple
test files), so not fringe.

2. Coverage matches by name only → false positives on duplicate names

`ruleMatchesSelector` and the alert-marking loop match purely on rule **name**, ignoring
group/file (unittest.go:329-340, 858). Two rules sharing a name in different groups are
both marked covered when only one is tested.

Repro: `SameName` alerts in `groupA`/`groupB` and `dup:metric` records in `recA`/`recB`,
each name tested once → `Rule test coverage: 4/4 rules covered (100.0%)` (and one of the
`SameName` tests actually *failed*).

LLM suggests :
Drop `loadRuleEntries` (~47 lines) and build entries from `groupsMap`:

```go
for _, g := range groupsMap {
    for _, r := range g.Rules() {
        e := ruleEntry{group: g.Name(), file: g.File()}
        switch rule := r.(type) {
        case *rules.AlertingRule:
            e.name, e.rtype, e.labels = rule.Name(), "alert", rule.Labels()
        case *rules.RecordingRule:
            e.name, e.rtype, e.labels = rule.Name(), "record", rule.Labels()
        }
        // collect, keyed by (file, group, name, type)
    }
}
```

File: cmd/promtool/unittest.go [hunk 15]
```diff
@@ -72,18 +80,22 @@ func RulesUnitTestResult(results io.Writer, queryOpts promqltest.LazyLoaderOpts,
...
 		}
 		fmt.Println()
+		allRuleEntries = append(allRuleEntries, entries...)
 	}
 	err := junit.WriteXML(results)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "failed to write JUnit XML: %s\n", err)
 	}
+	if coverage && len(allRuleEntries) > 0 {
+		printCoverageSummary(allRuleEntries)
```
Comment (krajorama) [inline cmd/promtool/unittest.go]: How can one fail CI if coverage isn't 100% ? This just prints , but doesn't alter the return code. Wouldn't your use case need another flag to force returning error on less than 100% coverage?

**Review by krajorama — COMMENTED** [review]
Thank you for the contribution. The keying on alert name was very suspicious, so I ran LLM on that aspect and also I felt that `loadRuleEntries` was duplicating a lot of code.  I've commented on these aspects.

File: cmd/promtool/unittest.go (hunk 13 above)
Comment (thc1006) [reply cmd/promtool/unittest.go]: Done. Introduced a `ruleKind` type with `alertingRule`/`recordingRule` constants and a `String()` method (returning `"alert"`/`"record"` for the report), and removed the copy-paste strings.

File: cmd/promtool/unittest.go (hunk 14 above)
Comment (thc1006) [reply cmd/promtool/unittest.go]: Done — thanks, this was the right call. `loadRuleEntries` is gone; coverage is now built by calling `rules.Manager.LoadGroups` (Prometheus's own loader, not a hand-rolled `rulefmt` parser) and inspecting the resulting `*rules.Group` objects, so it reuses Prometheus's group+rule label merge. Rules are keyed by `(file, group, name, kind, labels, query)` in a single suite-wide registry.

Both issues are addressed, with regression tests:

1. **Double-counting across test files** — the shared registry de-duplicates by the key above (file path canonicalized with `filepath.Abs`), so a rule file referenced by N test files is counted once. See `TestRuleCoverageDeduplicatesAcrossFiles`.
2. **Name-collision false positives** — rules are attributed to their actual loaded group object, so same-named rules in different groups are distinct in the denominator, and recording rules are further disambiguated by static-label compatibility. See `TestRuleCoverageDistinguishesDuplicateNames`.

One honest nuance worth flagging: any two **alerting** rules that share an `alertname` across groups — regardless of their labels or expressions — are both marked covered by a single name-based `alert_rule_test`, because promtool evaluates same-named alerts together and the assertion compares against their unioned output (per the unit-testing docs). They're counted as two distinct rules in the denominator but both marked covered. Recording rules are disambiguated by their static labels. I've documented this and the other boundaries (it can also under-report on `__name__`-less selectors; a rule whose only test fails still counts as "covered"; `--run` is honored) in the PR description rather than overstating precision.

(I've also rebased the branch onto current `main` and re-verified build, tests, `-race`, `golangci-lint`, and `make doc` against it.)

File: cmd/promtool/unittest.go (hunk 15 above)
Comment (thc1006) [reply cmd/promtool/unittest.go]: Added `--coverage-threshold=<0-100>` (float, default `0` = off, following the coverage.py / go-test-coverage convention). When set it implies `--coverage` and exits `1` if total coverage is below the threshold, so CI can gate on it; otherwise `--coverage` is report-only. The gate compares the percentage exactly as displayed (one decimal place) so the exit code never contradicts the printed number. Covered by `TestRuleCoverageThresholdValidation` / `TestRuleCoverageThresholdRounding` / `TestBelowThreshold`.

**Comment by thc1006** [comment]
@dgl sorry for the late reply. On #16887 you ask for a small design document about how the coverage can actually work, and I go to the code directly. That is my mistake.

The problem you point out there is fixed now. The selectors come from `ExtractSelectors`, so a test that writes `sum by (job) (job:up:sum)` still counts for the recording rule, and the rule labels are compared too, not only the name.

For the design part I put it in the PR description. It explains what "covered" means here, why I not use the input-feeding idea from #8131, how the rules are identified, where the attribution is wrong in both direction, and what `--coverage-threshold` can promise and what it cannot.

So my question is: the PR description is enough for you, or you prefer I write a proposal in prometheus/proposals first? Both way is fine for me. If proposal is better, I want to start it before other people spend more time on the review here. No hurry, please take your time.

@krajorama thank you very much for the review in June. All three comments are done: the `ruleKind` type, build from `LoadGroups` instead of a second loader, and `--coverage-threshold` so the CI can really fail.
